### ISSUE TITLE
Help and improvements 

### ISSUE THREAD

**Issue by rsdev000** [rinvex/repository#15]
Hi Abdelrahman Omran. Yesterday didn't had enough time to test your package and posted an issue based on code read only. Didn't mean to be rude and I'm sorry if the post sounds like it. I really appreciate your work.

Today got some time and started with a fresh laravel install and a simple database I use for development: Posts, Comments and Authors. Really simple structure with dummy data.

This is what I have and the errors that I found. Perhaps I am making some mistakes here but just followed your docs.

PostRepository
```php
<?php

namespace App\Repositories;

use App\Models\Post;
use Illuminate\Container\Container as Application;
use Rinvex\Repository\Repositories\EloquentRepository;

class PostRepository extends EloquentRepository
{
    // Instantiate repository object with required data
    public function __construct(Application $app)
    {
        $this->setContainer($app)
             ->retrieveModel(Post::class)
             ->setRepositoryId('rinvex.repository');
    }

    public function getPost10()
    {
        return $this->findWhere(['post_id', '<=', 10]);
    }
}
```

IndexController
```php
<?php

namespace App\Http\Controllers;

use App\Http\Requests;
use App\Repositories\PostRepository;

class IndexController extends Controller
{
    private $repository;

    public function __construct(PostRepository $repository)
    {
        $this->repository = $repository;
    }

    public function index()
    {
        $post20 = $this->repository->getPost10();
        return $post20;
    }
}
```

The application throws an error:

```
BadMethodCallException in Builder.php line 2345:
Call to undefined method Illuminate\Database\Query\Builder::setRepositoryId()

in Builder.php line 2345
at Builder->__call('setRepositoryId', array('rinvex.repository'))
at call_user_func_array(array(object(Builder), 'setRepositoryId'), array('rinvex.repository')) in Builder.php line 1402
at Builder->__call('setRepositoryId', array('rinvex.repository'))
at call_user_func_array(array(object(Builder), 'setRepositoryId'), array('rinvex.repository')) in Model.php line 3505
at Model->__call('setRepositoryId', array('rinvex.repository')) in PostRepository.php line 16
at PostRepository->__construct(object(Application))
at ReflectionClass->newInstanceArgs(array(object(Application))) in Container.php line 779
at Container->build('App\Repositories\PostRepository', array()) in Container.php line 629
at Container->make('App\Repositories\PostRepository', array()) in Application.php line 697
at Application->make('App\Repositories\PostRepository') in Container.php line 849
at Container->resolveClass(object(ReflectionParameter)) in Container.php line 804
at Container->getDependencies(array(object(ReflectionParameter)), array()) in Container.php line 773
...
```

Then read about Automatic Guessing in the docs and commented the __construct() method in PostRepository.php

Another error
```
FatalThrowableError in EloquentRepository.php line 150:
Call to a member function toSql() on null

in EloquentRepository.php line 150
at EloquentRepository->findWhere(array('post_id', '<=', '10')) in PostRepository.php line 21
at PostRepository->getPost10() in IndexController.php line 19
at IndexController->index()

Line 150: $cacheKey = md5(json_encode([$where, $columns, $with, $lifetime, $driver, $this->model->toSql()]));

$this->model is null
```

Can you give me some feedback? I would like to test the improvements on the caching system.
Thanks in advance.

Cheers

**Comment by Omranic** [comment]
Hello Dear @rsdev000,
First of all, I'm really thankful for your feedback, it's constructive & pushes the development forward 👍 That's always welcome and encouraged, please keep doing it 😉 

This time, it's my fault. There's a mistake in the docs, the `retrieveModel` method must be the last called method in the constructor like that:
```php
$this->setContainer($app)
        ->setRepositoryId('rinvex.repository')
        ->retrieveModel(Post::class);
```

I'll update the docs to fix this, everything else is fine in your example. Give it another shot accordingly, and let me know your results 😃

**Comment by rsdev000** [comment]
It's working now :) . I'll continue digging in the package and provide you with all feedback that I can.

Cheers

**Comment by Omranic** [comment]
Always welcome 😊

**Comment by rsdev000** [comment]
I'm making a mistake here. 
return $this->findWhere(['post_id', '<=', 10]); is returning all posts and I just want all posts with post_id <= 10.

Another documentation improvement ? :)

**Comment by Omranic** [comment]
Yeah, seems to be another docs issue 😄 
I'll check this & update you soon..

**Comment by rsdev000** [comment]
Can you advance the correct syntax while fixing the docs ?

**Comment by Omranic** [comment]
Look like a bug, not just docs syntax, working on it.. :)

**Comment by rsdev000** [comment]
Take your time :)

**Comment by Omranic** [comment]
OK, here's a quick fix until I release the update since I'm implementing a new feature as well #16 

In the `EloquentRepository` class, modify the following lines from ([source](https://github.com/rinvex/repository/blob/master/src/Repositories/EloquentRepository.php#L156-L158)):
```php
                if (is_array($value)) {
                    list($attribute, $condition, $value) = $value;
                    $this->model->where($attribute, $condition, $value);
                } else {
                    $this->model->where($attribute, '=', $value);
                }
```

To:
```php
                if (is_array($value)) {
                    list($attribute, $condition, $value) = $value;
                    $this->model = $this->model->where($attribute, $condition, $value);
                } else {
                    $this->model = $this->model->where($attribute, '=', $value);
                }
```

**Updated Syntax Docs:**
```php
// Matching values with equal '=' operator
$result = $repository->findWhere(['id' => 1]);

// Same way you can add multiple where conditions
$result = $repository->findWhere(['id' => 1, 'slug' => 'example']);

// Matching exact value with custom operator (notice that parameter syntax differs this time, it's now array of arrays)
$result = $repository->findWhere([['id', '=', 1]]);

// Same way you can add multiple where conditions
$result = $repository->findWhere([['id', '=', 1], ['slug', '!=', 'example']]);
```

**Comment by rsdev000** [comment]
Working as expected :)

What about implementing a method that returns a Builder instance and use it fluently?
The main issue could be caching the results.

**Comment by Omranic** [comment]
All reported issues now solved, and a new update released `v1.0.2`.

In regard to returning a builder instance, I don't think that's within the scope of this package, but you still have the ability to add whatever functionality in your own repository. 

Keep up the good work 👍
